src/main/java/seedu/exercise/logic/commands/ClearCommand.java

```diff
@@ -2,22 +2,31 @@
 
 import static java.util.Objects.requireNonNull;
 
+import seedu.exercise.logic.commands.history.EventHistory;
 import seedu.exercise.model.ExerciseBook;
 import seedu.exercise.model.Model;
+import seedu.exercise.model.ReadOnlyExerciseBook;
 
 /**
  * Clears the exercise book.
  */
-public class ClearCommand extends Command {
+public class ClearCommand extends Command implements UndoableCommand {
 
     public static final String COMMAND_WORD = "clear";
     public static final String MESSAGE_SUCCESS = "Exercise book has been cleared!";
 
+    private ReadOnlyExerciseBook exerciseBookPrevious;
 
     @Override
     public CommandResult execute(Model model) {
         requireNonNull(model);
+        exerciseBookPrevious = new ExerciseBook(model.getAllExerciseData()); // Create a deep copy
         model.setExerciseBook(new ExerciseBook());
+        EventHistory.getInstance().addCommandToUndoStack(this);
         return new CommandResult(MESSAGE_SUCCESS);
     }
+
+    public ReadOnlyExerciseBook getExerciseBookPrevious() {
+        return exerciseBookPrevious;
+    }
```

Comment:
Should the command itself keep the previous exercise book? Seems like it violates SRP since ClearCommand sole responsibility is to execute the clear command by the user.

Maybe move the `exerciseBookPrevious` into the `EventHistory` since it is used there

Reply:
Same for the remaining undoable commands

Reply:
I have decided to stick to the current design because I am truly unable to find simpler solutions that would not introduce unnecessary and unjustifiable dependencies for EventHistory and EventFactory classes.

To justify the storage of the previous exercise book in ClearCommand, I am renaming the variable from ExerciseBookPrevious to ExerciseBookCleared to indicate that this ExerciseBookCleared is unique to the ClearCommand, like how the ExerciseDeleted is unique to the DeleteCommand. This does make a bit of sense in showing that this ClearCommand was particularly involved in clearing this version of the exercise book.

TLDR: Really need to store the older version of ExerciseBook in ClearCommand to access it in the EventFactory when producing the ClearEvent.